File: exercism/ocaml/custom-set/custom_set.ml
open Core.Std

module type ITEM = sig
  type t
  val compare: t -> t -> int
  val equal: t -> t -> bool
  val to_string: t -> string
end

module Make(T: ITEM) = struct
  type t = T.t list

  let empty : t = []

  let equal : t -> t -> bool = (=)

  let to_string (set: t) : string =
    "{" ^ String.concat ~sep:" " (List.map ~f:T.to_string set) ^ "}"

  let rec compare (set1: t) (set2: t) : int = match set1, set2 with
  | [], [] -> 0
  | [], _ -> -1
  | _, [] -> 1
  | hd1 :: tl1, hd2 :: tl2 -> match T.compare hd1 hd2 with
    | 0 -> compare tl1 tl2
    | i -> i

  let rec add (set: t) (item: T.t) : t = match set with
  | [] -> [item]
  | x :: xs -> match T.compare item x with
    | n when n < 0 -> item :: set
    | 0 -> set
    | _ -> x :: add xs item
    
  let of_list : T.t list -> t =
    List.fold ~init:empty ~f:add

  let to_list : t -> T.t list = Fn.id

  let union (set1: t) (set2: t) : t =
    List.fold ~init:set1 ~f:add set2

  let rec remove (set: t) (item: T.t) : t = match set with
  | [] -> []
  | x :: xs -> match T.compare item x with
    | n when n > 0 -> x :: remove xs item
    | 0 -> xs
    | _ -> set

  let difference (set1: t) (set2: t) : t =
    List.fold ~init:set1 ~f:remove set2

  let intersect (set1: t) (set2: t) : t =
    set2 |> difference set1
         |> difference set1

end

File: exercism/elixir/custom-set/custom_set.exs
defmodule CustomSet do

  @opaque t :: %__MODULE__ { content: Map.t }
  defstruct content: %{}

  @spec new(Enum.t) :: t
  def new(content) do
    %CustomSet { content: Map.new(content, fn k -> {k, true} end) }
  end

  @spec empty?(t) :: boolean
  def empty?(set) do
    Enum.empty?(set.content)
  end

  @spec contains?(t, any) :: boolean
  def contains?(set, item) do
    Map.has_key?(set.content, item)
  end

  @spec subset?(t, t) :: boolean
  def subset?(subset, superset) do
    empty?(difference(subset, superset))
  end

  @spec disjoint?(t, t) :: boolean
  def disjoint?(set1, set2) do
    equal?(difference(set1, set2), set1)
  end

  @spec equal?(t, t) :: boolean
  def equal?(set1, set2) do
    Enum.sort(set1.content) == Enum.sort(set2.content)
  end

  @spec add(t, any) :: t
  def add(set, item) do
    %__MODULE__ { content: Map.put(set.content, item, true) }
  end

  @spec intersection(t, t) :: t
  def intersection(set1, set2) do
    difference(set1, difference(set1, set2))
  end

  @spec difference(t, t) :: t
  def difference(set1, set2) do
    %__MODULE__ { content: Map.drop(set1.content, Map.keys(set2.content)) }
  end

  @spec union(t, t) :: t
  def union(set1, set2) do
    %__MODULE__ { content: Map.merge(set1.content, set2.content) }
  end

  @spec size(CustomSet.t) :: non_neg_integer
  def size(set) do
    length(set.content)
  end
  
  @spec to_list(CustomSet.t) :: List.T
  def to_list(set) do
    Enum.sort(set.content)
  end
end
